lib/saluki-components/src/common/otlp/mod.rs

```diff
@@ -0,0 +1,207 @@
+                .route("/v1/traces", post(http_traces_handler::<H>))
+                .with_state((http_handler, memory_limiter.clone())),
+        );
+
+        let http_listener = ConnectionOrientedListener::from_listen_address(self.http_endpoint)
+            .await
+            .map_err(|e| generic_error!("Failed to create OTLP HTTP listener: {}", e))?;
+
+        let http_server = HttpServer::from_listener(http_listener, service);
+        let (http_shutdown, http_error) = http_server.listen();
+
+        Ok((http_shutdown, http_error))
+    }
+}
+
+/// HTTP handler for OTLP metrics requests.
+async fn http_metrics_handler<H: OtlpHandler>(
+    State((handler, memory_limiter)): State<(std::sync::Arc<H>, MemoryLimiter)>, body: Bytes,
+) -> (StatusCode, &'static str) {
+    memory_limiter.wait_for_capacity().await;
+
+    match handler.handle_metrics(body).await {
+        Ok(()) => (StatusCode::OK, "OK"),
+        Err(msg) => {
+            error!("Failed to handle OTLP metrics: {}", msg);
+            (StatusCode::INTERNAL_SERVER_ERROR, "Internal processing error")
+        }
+    }
+}
+
+/// HTTP handler for OTLP logs requests.
+async fn http_logs_handler<H: OtlpHandler>(
+    State((handler, memory_limiter)): State<(std::sync::Arc<H>, MemoryLimiter)>, body: Bytes,
+) -> (StatusCode, &'static str) {
+    memory_limiter.wait_for_capacity().await;
+
+    match handler.handle_logs(body).await {
+        Ok(()) => (StatusCode::OK, "OK"),
+        Err(msg) => {
+            error!("Failed to handle OTLP logs: {}", msg);
+            (StatusCode::INTERNAL_SERVER_ERROR, "Internal processing error")
+        }
+    }
+}
+
+/// HTTP handler for OTLP traces requests.
+async fn http_traces_handler<H: OtlpHandler>(
+    State((handler, memory_limiter)): State<(std::sync::Arc<H>, MemoryLimiter)>, body: Bytes,
+) -> (StatusCode, &'static str) {
+    memory_limiter.wait_for_capacity().await;
+
+    match handler.handle_traces(body).await {
+        Ok(()) => (StatusCode::OK, "OK"),
+        Err(msg) => {
+            error!("Failed to handle OTLP traces: {}", msg);
+            (StatusCode::INTERNAL_SERVER_ERROR, "Internal processing error")
+        }
+    }
+}
+
+/// gRPC service implementation that delegates to the handler.
+struct GrpcServiceImpl<H> {
+    handler: std::sync::Arc<H>,
+    memory_limiter: MemoryLimiter,
+}
+
+impl<H> GrpcServiceImpl<H> {
+    fn new(handler: std::sync::Arc<H>, memory_limiter: MemoryLimiter) -> Self {
+        Self {
+            handler,
+            memory_limiter,
+        }
+    }
+}
+
+#[async_trait]
+impl<H: OtlpHandler> MetricsService for GrpcServiceImpl<H> {
+    async fn export(
+        &self, request: TonicRequest<ExportMetricsServiceRequest>,
+    ) -> Result<Response<ExportMetricsServiceResponse>, Status> {
+        self.memory_limiter.wait_for_capacity().await;
+
+        let raw_bytes = request.into_inner().encode_to_vec();
+
+        match self.handler.handle_metrics(Bytes::from(raw_bytes)).await {
```

Comment:
This is fine for now, but since this is ultimately riding over HTTP... we probably want to explore trying to just manually set up the HTTP handlers so that we get the raw Protocol Buffers payloads without actually decoding.

This would let us avoid having to decode and then re-encode just to get back to `Bytes`.